doc/rst/technotes/optimization.rst

```diff
@@ -0,0 +1,662 @@
+``--no-local``. You are seeing wide pointer overhead if the ``--local``
+version is significantly faster than the ``--no-local`` version.
+
+What can be done about it?
+
+ * use ``local`` blocks to tell the compiler that code within a block
+   will not communicate. That allows it to remove wide pointers for that
+   code.
+ * use ``localAccess`` to access local elements of an array
+ * use ``localSubdomain`` to compute the local index set as a local domain
+ * use ``localSlice`` to compute a non-distributed array slice that refers
+   to the local portion
+
+Fine-Grained Communication
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+It's easy to write Chapel programs that use fine-grained communication.
+Such programs will work with many small messages and the performance will
+be sensitive to the latency of a message in the network. To make such
+programs perform and scale better, try to get the communication to use
+larger messages so that the network is operating more in a
+bandwidth-bound way rather than a latency-bound way.
+
+Let's look at an example of permuting an array. Suppose that we have
+three distributed arrays all over the distributed domain ``D``:
+
+ * ``In`` stores input values
+
+ * ``Idx`` stores a permutation; each index in ``D`` occurs exactly once
+   in ``Idx``.
+
+ * ``Out`` stores the result of permuting ``In`` according to ``Idx``
+
+In particular, we will set ``Out[Idx[i]] = In[i];`` for each ``i`` in ``D``.
+
+If we write that in the simplest way, we get this code:
+
+.. literalinclude:: ../../../test/technotes/optimization/aggregation.chpl
+   :language: chapel
+   :start-after: START_UNOPT
+   :end-before: END_UNOPT
+
+However that code will do fine-grained PUT operations on the network;
+``Out[Idx[i]]`` will refer to the array element to store into, and since
+that array element is typically stored on a different locale, the
+compiler will generate a PUT operation for ``Out[Idx[i]] = In[i]``.
+
+We can optimize this by using the :chpl:mod:`CopyAggregation` module.
+Here is the optimized version:
+
+.. literalinclude:: ../../../test/technotes/optimization/aggregation.chpl
+   :language: chapel
+   :start-after: START_OPT
+   :end-before: END_OPT
+
+In this case, the optimization required two changes to the program:
+
+ 1. Create a copy aggregator per-task. Here, we use the ``forall`` intent
+    ``with`` clause to create a ``DstAggregator`` per task. The
+    ``DstAggregator`` will optimize the PUT operations by combining PUT
+    operations destined for the same locale.
+
+ 2. Use the copy aggregator instead of plain old assignment. The call
+    ``agg.copy(X, Y)`` is functionally similar to ``X = Y``.
+
+How much does optimizing this example improve performance? With a quick
+test on 16 nodes, we can see (through use of the CommDiagnostics module)
+that the number of PUTs is approximately 1000 times less with the
+optimized version, and the performance is approximately 30 times better.
```

Comment:
Should you mention the auto-aggregation flag here? Would it work on this example?